notes/fomega/evaluation/builtin-evaluation.md

```diff
@@ -0,0 +1,285 @@
+exactly how builtin evaluation works, but we'll need to be more
+precise (and independent of the implementation language) if and when
+we do away with static builtins.  Maybe the machine implementation
+should be one that's easily specified.  An argument in favour of the
+CK machine was its simplicity, but maybe it's not possible to be very
+simple if we deal with builtins properly.
+
+Also, our primary evaluator will eventually be untyped, and it'll
+presumably be a CEK machine.  We'll surely still want a typed
+evaluator as well though, so maybe that should be the CEK machine as
+well.  It's maybe tempting to use the typed CEK machine and make it a
+bit more robust by using the `TypeArg | TermArg` thing mentioned above
+to count the arities: that'll be slightly less efficient, but we'll
+only need it for the typed evaluator, where efficiency isn't paramount.
+
+
+## Higher-rank builtins
+
+In the specification, builtins are currently allowed to be
+polymorphic, but all the type arguments have to precede the term
+arguments (ie, all the quantifiers appear at the front).  
+
+We've recently had a discussion about whether we want to allow
+type and term arguments to be interleaved, allowing us to
+have builtins with higher-rank types (and also non-canonical
+rank-one types).  The internal machinery doesn't really care
+about this, but it does present some problems.
+
+One possible issue is that if we allow type arguments between term
+arguments there's the question of how we erase them.  If we allow such
+things, then we can have a polymorphic partially-applied builtin, and
+such a thing could be passed to a function which could also accept
+type-abstracted lambdas, and such a function would have to deal
+uniformly with both possibilities.  The difficulty is that we no longer
+have the value restriction, so we have to erase abstractions to
+`delay` and apply `force` when type instantiation occurs.  Since
+there's no possibility of computation occurring when a
+partially-applied builtin is instantiated, it seems strange to
+unnecessarily wrap them in a `delay`. [But what happens if the final
+argument is a type argument, ie, if a builtin returns a polymorphic
+value?  We currently require all builtin arguments to be values and
+builtins are supposed to be parametrically polymorphic _and_ return
+values, so can a builtin really return something that can compute when
+instantiated?  Perhaps it can: maybe we can pass `Λt:*.error` (which
+erases to `delay error`) to the identity function?]
+
+Anyway, perhaps we could modify `force` so that it'll act on _any_
+value, doing nothing when it's not wrapped in `delay`; we would then
+be able to completely erase type arguments to builtins except perhaps
+ones at the very end.  (I'm now wondering what would happen if you
+force a _term_: could that ever happen?)
+
+Roman has a number of other proposals, including erasing type arguments
+of builtins to `()` and prohibiting final quantifiers altogether.
```

Comment:
Now I'm not sure exactly what we mean by a final quantifier.  Are we talking about the case when the final argument is a type argument, so you supply all the term arguments and then you still have to instantiate the builtin before in computes?  I'm finding it hard to see why that would occur.  It's a little confusing because the return type could be a forall, and then you'd have to instantiate the _result_ of the builtin; I don't that's what we're talking about here though.

Reply:
Well, consider an error builtin, with type `forall a. a`. This takes precisely one argument, a type argument, and only then evaluates to error. Seems at least somewhat reasonable?

Reply:
No, I think not. The CAM doesn't know anything about types: you give it a builtin and a bunch of terms and it computes a result and returns it.  It doesn't know anything about types, so it computes at the point when it gets its final term argument.  Now you could return a polymorphic value that's just like any other polymorphic value: you can instantiate it, and at that point it might carry out some more computation _outside the constant application machinery_.  There is a quantifier at the end, but it's in the _return type_ of the builtin.  This sort of thing always confuses me because of the fact that builtins have type schemes that don't correspond one-to-one with types: when you convert a type scheme to a type you just get lots of arrows and if you don't know the arity you can't tell whether the arrows at the end belong to the return type or the input type.  

I think something similar is going on with builtins.  If you have a builtin b with signature `[t₁,...,tₙ]r` and we're talking about a final quantifier, does it belong in `tₙ` or in `r`?  If it's in `r` then b's just returning a polymorphic value (which is perfectly legitimate: b might be the identity function and you've given it an abstraction) and there's no problem about instantiating it.  If the quantifier's in `tₙ`, then what have we got?  I think `tₙ` has to be `∀t:K...`, in which case the final argument's a polymorphic value and it doesn't make sense to talk about instantiating the application. 

Does this make sense, or am I misunderstanding?  If we consider a signature like `[t₁,...,tₙ]r`, how are we interpreting the `t`s?  I'm not sure how to make sense of saying that some t represents a type argument (as opposed to a term). If we're saying that a builtin can take an actual type as an argument, then I don't see how that could work: the CAM ignores types, so they can't have any effect on what goes on inside the builtin, and they can't somehow delay evaluation because nothing happens anyway until you get the final _term_ argument. Even if the final argument was a type, wouldn't the builtin still have to compute when it gets the final term?  If there's any delay, it would have to be handled by the machine, not the CAM.

[Off to lie down: I've got no idea what's going on any more.]

Reply:
> No, I think not. The CAM doesn't know anything about types: you give it a builtin and a bunch of terms and it computes a result and returns it. It doesn't know anything about types, so it computes at the point when it gets its final term argument.

This seems wrong if we're serious about turning type arguments into delays that require a force to get past them. If the last argument isn't a type, then you can get away with just ignoring any forces that are applied (assuming they go in the right place), because in the end you'll have to get all your term arguments. Having a type argument as the last argument just puts this into focus.

So I think the CAM *should* consider type arguments.

(I agree that there's a difference between the quantifier being at the end and it being in the return type. But I think those two things *just are different*. And I don't see why the version where the quantifier is at the end of the type argument list shouldn't work.)

Reply:
@kwxm 

> Are we talking about the case when the final argument is a type argument, so you supply all the term arguments and then you still have to instantiate the builtin before in computes?

Ah, sorry, yes, I meant that + silently removing all types from builtin applications (which is what you've convinced me was a bad idea). Prohibiting final quantifiers without silently removing types (but using explicit `force` for example) could work indeed.

@michaelpj 

> Well, consider an error builtin, with type forall a. a. This takes precisely one argument, a type argument, and only then evaluates to error. Seems at least somewhat reasonable?

If we ever have polymorphic built-in types, then also `nil : all a. list a`.

Reply:
> silently removing all types from builtin applications (which is what you've convinced me was a bad idea).

I'm becoming convinced that it's a good idea!

Reply:
> So I think the CAM should consider type arguments.

That disturbs me.  I don't like it when type-level constructs influence computation.  I tend to feel that you should just be able to throw all the types away without changing anything, but maybe that's just me being narrow-minded.

Reply:
>That disturbs me. I don't like it when type-level constructs influence computation. I tend to feel that you should just be able to throw all the types away without changing anything, but maybe that's just me being narrow-minded.

I think we've already gone down this road with erasing big lambda/type instantiation to delay/force so we should keep things consistent.